default/AI/ShipDesignAI.py

```diff
@@ -1308,6 +1256,28 @@ def _calc_rating_for_name(self):
         self.update_stats(ignore_species=True)
         return self.structure
 
+    def _adjusted_production_cost(self):
+        """Return a production cost adjusted by the number of ships we have.
+
+        Building one warship of rating 2R and cost 2C is effectively cheaper than building 2 warships of rating R and
+        cost C due to fleet upkeep considerations even if they have the same raw rating/cost ratio.
+        The significance of this fleet upkeep efficiency increases with the raw fleet upkeep rate, with the expected
+        lifetime of the ship under consideration (i.e., with a longer expected lifespan it will affect the future
+        construction of a greater number of ships), and it also increases with the number of ships of this design
+        expected to be created. In the calculation below, the empire's current total shipcount is taken as a rough proxy
+        for both of the duration and extent factors.
+        Note: This same sort of adjustment would be valid for troop ships, could theoretically have some applicability
+        to scout ships since there could conceivably be a tradeoff for more scout ships of lower rating/range, but
+        would really not be applicable to colony ships since in that case there is really not a potential tradeoff
+        between number of ships and capacity.
+
+        :return: adjusted production cost
+        :rtype: float
+        """
+        # TODO: Consider total pp production output as additional factor
```

Comment:
I was not talking about the cost cache but the cache used when designing the ships. As the PP output per turn is a quantity that can easily change every single turn, the design cache will be pretty much turn-based only. Maybe that is no problem though, performance definitely improved since I first added the cache.

Another thing: I do not think that ```cost/turn``` is a good quantity to determine which ships we can build. Instead, we should only consider total PP cost. Consider the following case:
Design A has a rating of 100, PP cost of 100 and turn time of 10.
Design B has a rating of 150, PP cost of 80 and turn time of 4.

So cost per turn for Design B is higher. But that is entirely meaningless - if we simply build the ship over the 10 turns (by swapping it around in the queue), we would obviously end up with a better ship in the same time for less cost compared to the case where we decide to build Design A.